# 例外処理
- rescue句
  - 引数は全て省略可能
  - begin ~ endの中に、複数のrescueを記述可能
- retry句: beginから処理をやり直す
- raise句: 例外を発生させる
```ruby
rescue [補足したいエラークラス, その2, ...] => エラークラスを受け取る変数

raise Errorクラス, 'エラーメッセージ'
```

# transaction
## 概要/構文
- ブロックの中で例外が起きたら、rollbackを発生させる
```ruby
begin
  Model.transaction do #ActiveRecord::Base.transaction doでも可
    instance.orm_method!
  end
rescue => error
  #例外処理
end
```

## テクニック
### 意図的に例外を起こす
- transaction内でraiseする、その際、エラーオブジェクトはActiveRecord::Rollbackが使われている記事が多い
  - raiseのみでもrollback可能なため、意図は可読性？
### ネストさせる
- 全く理解できていない


# logファイルへの出力
- config/environments/実行環境.rb を編集
- log/ にlogファイルを用意
```ruby
# config/environments/実行環境.rb
config.メソッド名 = Logger.new(Rails.root.join('log/logファイル'))

# logファイルへ書き込み
Rails.application.config.メソッド名.info("出力テキスト")
```
